### DISCUSSION TITLE
Possibility to install D8VK and maybe others for Heroic?

### DISCUSSION THREAD

**Discussion by stuart-bowyer** [DavidoTek/ProtonUp-Qt#208] [Ideas]
I use Heroic over Lutris because it just seems like a better experience. Any chance to add D8VK and maybe other compatibility tools to Heroic? Thanks and keep up the excellent work!

**Comment by begin-theadventure** [comment]
D8VK will be merged with DXVK in the future, for now just download D8VK release and drop the files into the DXVK folder that Heroic uses.

**Reply by stuart-bowyer** [reply]
Thanks this is helpful for the time being, but it seems like DavidoTek might implement this in the future too.
Thanks!

**Comment by DavidoTek** [comment]
Yes, it should be possible to install D8VK and DXVK for Heroic.

I will need to include the paths for Heroic in ProtonUp-Qt: `~/.var/app/com.heroicgameslauncher.hgl/config/heroic/tools/dxvk/` and `~/.config/heroic/tools/dxvk/`

**Reply by stuart-bowyer** [reply]
Hi DavidoTek, thanks very much! Heroic seems quite good for older PC games.It just feels more polished than Lutris and cleaner with the prefixes etc in my opinion.

Thanks very much again and I look forward to seeing this in the future.

**Reply by sonic2kk** [reply]
Just a note that the D8VK and DXVK install directories are somewhat-hardcoded for Lutris. They point to `../../runtime/dxvk`.

It should be possible o check the `install_dir` and infer how we're installing it that way, though it would keep the ctmod cleaner if we passed the install directory to these ctmods. It should be possible to do, but would require a bit of a refactor I think.

I'd also like to mention that Heroic supports vkd3d as well, so we could include support for that. Just like with DXVK though, the vkd3d ctmods point to the Lutris install directories, so the same sort of refactor would be required.

<hr>

As for the paths, Heroic can also be used from AppImage/package manager, and the path in this case is `~/.config/heroic/` as opposed to the `~/.var/app/<id>` path (it doesn't use `~/.local/share/heroic`, I just ran the AppImage to verify). But looking at the Flatpak paths, ~~it seems like ProtonUp-Qt [may not have any paths for non-Flatpak Heroic](https://github.com/flathub/net.davidotek.pupgui2/blob/master/net.davidotek.pupgui2.json#L25) at the moment?~~ ProtonUp-Qt just has general access to `~/.config`.

**Reply by DavidoTek** [reply]
> it would keep the ctmod cleaner if we passed the install directory to these ctmods. It should be possible to do, but would require a bit of a refactor I think.

Yeah. I already thought about that a bit. Ideally, the [launcher definitions in `constants.py`](https://github.com/DavidoTek/ProtonUp-Qt/blob/f73f9bd04d3b8c7d0d03bbe3b83f20331debd9fc/pupgui2/constants.py#L33) would not include just one `install_dir`, but rather a list(/dict/class) of install directories, e.g.
```python
{
    'install_dirs': {
        'proton': '~/.config/heroic/tools/proton/',
        'wine': '~/.config/heroic/tools/wine/',
        'dxvk': '~/.config/heroic/tools/dxvk/'
    },
    'display_name': 'Heroic',
    ...
}
```

That would make the `constants.py` quite messy though, maybe there is a cleaner approach? Either some sort of JSON file or alternatively a folder with one (Python script) file for each launcher, like the ctmods.

Then, the ctmods would be loaded by launcher, but rather by type definition:
```diff
- CT_LAUNCHERS = ['steam', 'heroicproton', 'bottles']
+ CT_TYPE = 'proton'
```

**Reply by sonic2kk** [reply]
A potential solution as described in #233 (to infer the launcher from the `install_dir` in the ctmod) would work. I tested this in the D8VK ctmod `get_tool` method, here is a snippet of the rough implementation:

```
dxvk_ext_dir = '../../runtime/dxvk'  # Default to Lutris path
if 'heroic/tools' in install_dir:
    dxvk_ext_dir = '../dxvk'

dxvk_dir = os.path.join(install_dir, dxvk_ext_dir)
```

If the `install_dir` contains Heroic, currently it is either `/path/to/heroic/tools/wine` or `/path/to/heroic/tools/proton`, so we can safely assume `../dxvk` is the correct path. And with this, D8VK will actually correctly install to the Heroic DXVK dir. The directory structure is sane enough that D8VK should be able to be loaded by Heroic (it matches the DXVK 2.1 I had with Heroic, except D8VK is 32bit only and doesn't have an `x64` dir). I didn't test but ProtonUp-Qt should be doing everything right up to this point.

The problem with this though: ProtonUp-Qt only lists tools visible in the current launcher's path. For Heroic this means it only lists tools in the Wine or Proton dirs. Which makes sense, but in this case it means it doesn't list tools in the other folders.

It would be possible maybe to merge the `heroicwine` and `heroicproton` into one launcher, but maybe the solution here is to add another launcher: `heroictools`. This could manage installing DXVK and vkd3d-proton, and we could filter out the Wine/Proton folders. Another similar solution would be to have an option somewhere for Heroic to allow tools and then display DXVK and vkd3d-proton in `heroicwine` and `heroicproton` (and if this option is disabled, maybe hide tools for Heroic? Not sure).

The reason I mention both `heroicwine` and `heroicproton` is that afaik Heroic lets you override the DXVK/vkd3d-proton version even if your version of Wine/Proton already comes with one. Also, while it is more common for a regular Wine build to not have these components, there are also cases where Proton may not have these either (such as kron4ek's builds). So I think it's useful to expose these tools for both Wine and Proton. The question is if we want to manage it this way or find another solution.

**Comment by DavidoTek** [comment]
Implemented with https://github.com/DavidoTek/ProtonUp-Qt/pull/245

It will be available in the next release.
